crates/host/src/policy.rs

```diff
@@ -260,18 +256,97 @@ impl Manager {
+            claims: claims.map(PolicyClaims::from),
+        };
+        self.evaluate_action(RequestBody::StartComponent(request))
+            .await
+    }
+
+    /// Use the policy manager to evaluate whether a provider may be started
+    #[instrument(level = "trace", skip_all)]
+    pub async fn evaluate_start_provider(
+        &self,
+        provider_id: impl AsRef<str>,
+        provider_ref: impl AsRef<str>,
+        annotations: &BTreeMap<String, String>,
+        claims: &Option<&jwt::Claims<jwt::CapabilityProvider>>,
+    ) -> anyhow::Result<Response> {
+        let request = ProviderInformation {
+            provider_id: provider_id.as_ref().to_string(),
+            image_ref: provider_ref.as_ref().to_string(),
+            annotations: annotations.clone(),
+            claims: claims.map(PolicyClaims::from),
+        };
+        self.evaluate_action(RequestBody::StartProvider(request))
+            .await
+    }
+
+    /// Use the policy manager to evaluate whether a component may be invoked
+    #[instrument(level = "trace", skip_all)]
+    pub async fn evaluate_perform_invocation(
         &self,
-        source: Option<RequestSource>,
-        target: RequestTarget,
-        action: Action,
+        component_id: impl AsRef<str>,
+        image_ref: impl AsRef<str>,
+        annotations: &BTreeMap<String, String>,
+        claims: &Option<&jwt::Claims<jwt::Actor>>,
+        interface: String,
+        function: String,
     ) -> anyhow::Result<Response> {
-        let cache_key = RequestKey {
-            source: source.clone().unwrap_or_default(),
-            target: target.clone(),
-            action: action.clone(),
+        let request = PerformInvocationRequest {
+            interface,
+            function,
+            target: ComponentInformation {
+                component_id: component_id.as_ref().to_string(),
+                image_ref: image_ref.as_ref().to_string(),
+                max_instances: 0,
+                annotations: annotations.clone(),
+                claims: claims.map(PolicyClaims::from),
+            },
+        };
+        self.evaluate_action(RequestBody::PerformInvocation(request))
+            .await
+    }
+
+    /// Sends a policy request to the policy server and caches the response
+    #[instrument(level = "trace", skip_all)]
+    pub async fn evaluate_action(&self, request: RequestBody) -> anyhow::Result<Response> {
+        let kind = match request {
+            RequestBody::StartComponent(_) => RequestKind::StartComponent,
+            RequestBody::StartProvider(_) => RequestKind::StartProvider,
+            RequestBody::PerformInvocation(_) => RequestKind::PerformInvocation,
+            RequestBody::Unknown => RequestKind::Unknown,
+        };
+        let cache_key = match request {
+            RequestBody::StartComponent(ref req) => RequestKey {
+                kind,
+                request: RequestBody::StartComponent(req.clone()),
```

Comment:
This is a lot of data to clone on every invocation. Anyway we can make RequestKey more lightweight by borrowing (or using Cows) instead?

Reply:
Let me see if we can drop it down to more required information

Reply:
Changed to, instead of cloning the full set of data, synthesizing a cache key based on a few of the strings. I also considered a Cow, or a use of lifetimes to reference a few strings for the cache key, but this is pretty simple and easy to understand. Let me know your thoughts or if I'm missing where `Cow` can make this a bit better